### PR TITLE
Prevent leaks when tasks fail - pulp_user

### PR DESCRIPTION
**Pull request by Alex-Welsh**

### CODE REVIEW

File: roles/pulp_user/tasks/main.yml [hunk 1]
```diff
@@ -18,37 +18,37 @@
```
Comment (markgoddard) [inline roles/pulp_user/tasks/main.yml:18]: L10 also loops over users

File: roles/pulp_user/tasks/main.yml [hunk 2]
```diff
@@ -18,37 +18,37 @@
 
 - name: Set users to delete fact
   set_fact:
-    remove_users: "{{ remove_users + [user.json.results[0]] }}"
-  loop: "{{ users_get_result.results }}"
+    remove_users: "{{ remove_users + [users_get_result.results[user_index].json.results[0]] }}"
+  loop: "{{ users_get_result.results | map(attribute='item') | map(attribute='username') }}"
```
Comment (markgoddard) [inline roles/pulp_user/tasks/main.yml:22]: I think you can do `attribute='item.username'`

File: roles/pulp_user/tasks/main.yml [hunk 3]
```diff
@@ -59,59 +59,63 @@
...
-    label: "{{ item.username }}"
+    index_var: user_index
   register: result
   changed_when: true
 
+- name: Set list of users that should exist
+  set_fact:
+    exist_users: "{{ create_users + update_users }}"
+
 - name: Add or remove user from group(s)
```
Comment (markgoddard) [inline roles/pulp_user/tasks/main.yml:102]: nit: You could use a task-scoped variable for `exist_users`

**Review by markgoddard — DISMISSED** [review]

File: roles/pulp_user/tasks/main.yml [hunk 4]
```diff
@@ -18,37 +20,52 @@
...
+      users_get_result.results[user_index].json.results[0].last_name != pulp_users[user_index].last_name) or
+      (pulp_users[user_index].email is defined and
+      users_get_result.results[user_index].json.results[0].email != pulp_users[user_index].email) or
+      (pulp_users[user_index].is_active is defined and
+      users_get_result.results[user_index].json.results[0].is_active != pulp_users[user_index].is_active) or
+      (pulp_users[user_index].is_staff is defined and
+      users_get_result.results[user_index].json.results[0].is_staff != pulp_users[user_index].is_staff) or
+      (pulp_users[user_index].groups is defined and
+      users_get_result.results[user_index].json.results[0].groups | map(attribute='name') | list != pulp_users[user_index].groups) or
+      pulp_users[user_index].password is defined
```
Comment (markgoddard) [inline roles/pulp_user/tasks/main.yml:68]: This is a bit susceptible to breaking when new parameters are added, but hopefully that won't happen too often.

**Review by markgoddard — APPROVED** [review]

File: roles/pulp_user/tasks/main.yml (hunk 4 above)
Comment (Alex-Welsh) [reply roles/pulp_user/tasks/main.yml:68]: Agreed. In general I'm not super happy with how it works but I think it's better to have it than not

**Pull request merged**